astronomy/date_body.hpp

```diff
@@ -872,32 +951,123 @@ constexpr std::array<int, (2017 - 1972) * 2> leap_seconds = {{
-  return date_time.time().is_end_of_day()
-             ? DateTimeAsUTC(date_time.normalized_end_of_day())
-             : CHECKING(
-                   !date_time.time().is_leap_second() ||
-                   (date_time.date().month() == 6 &&
-                    leap_seconds[(date_time.date().year() - 1972) * 2] == +1) ||
-                   (date_time.date().month() == 12 &&
-                    leap_seconds[(date_time.date().year() - 1972) * 2 + 1] ==
-                        +1),
-                   DateTimeAsTAIUnchecked(date_time) -
-                       UTC_TAI(date_time.date()));
+                   ModernUTC_TAI(Date::Calendar(utc_date.year(), 1, 1));
+}
+
+constexpr bool IsValidModernUTC(DateTime const& date_time) {
+  return !date_time.time().is_leap_second() ||
+         (date_time.date().month() == 6 &&
+          leap_seconds[(date_time.date().year() - 1972) * 2] == +1) ||
+         (date_time.date().month() == 12 &&
+          leap_seconds[(date_time.date().year() - 1972) * 2 + 1] == +1);
+}
+
+// Utilities for stretchy UTC (pre-1972).  This timescale includes rate changes
+// as well as fractional second leaps.
+
+// The (MJD - d) * t term from
+// https://hpiers.obspm.fr/iers/bul/bulc/UTC-TAI.history.
+constexpr quantities::Time RateTermTAI_UTC(DateTime const& utc) {
+  return utc.date() < "1962-01-01"_Date
+             ? (mjd(TimeScale(utc)) - 37'300) * 0.001'296 * Second
+       : utc.date() < "1964-01-01"_Date
+             ? (mjd(TimeScale(utc)) - 37'665) * 0.001'123'2 * Second
+       : utc.date() < "1966-01-01"_Date
+             ? (mjd(TimeScale(utc)) - 38'761) * 0.001'296 * Second
+             : (mjd(TimeScale(utc)) - 39'126) * 0.002'592 * Second;
+}
+
+// The constant term.
+constexpr quantities::Time OffsetTAI_UTC(Date const& utc_date) {
+  return utc_date < "1961-08-01"_Date ? 1.422'818'0 * Second
+       : utc_date < "1962-01-01"_Date ? 1.372'818'0 * Second
+       : utc_date < "1963-11-01"_Date ? 1.845'858'0 * Second
+       : utc_date < "1964-01-01"_Date ? 1.945'858'0 * Second
+       : utc_date < "1964-04-01"_Date ? 3.240'130'0 * Second
+       : utc_date < "1964-09-01"_Date ? 3.340'130'0 * Second
+       : utc_date < "1965-01-01"_Date ? 3.440'130'0 * Second
+       : utc_date < "1965-03-01"_Date ? 3.540'130'0 * Second
+       : utc_date < "1965-07-01"_Date ? 3.640'130'0 * Second
+       : utc_date < "1965-09-01"_Date ? 3.740'130'0 * Second
+       : utc_date < "1966-01-01"_Date ? 3.840'130'0 * Second
+       : utc_date < "1968-02-01"_Date ? 4.313'170'0 * Second
+                                      : 4.213'170'0 * Second;
+}
+
+// Returns TAI - UTC at the given point on the UTC timescale.
+constexpr quantities::Time StretchyTAI_UTC(DateTime const& utc) {
+  return OffsetTAI_UTC(utc.date()) + RateTermTAI_UTC(utc);
+}
+
+// Returns |true| if |utc| is within a leap of the given number of
+// |milliseconds| inserted before |next_day|.
+constexpr bool IsValidPre1972PositiveUTCLeap(DateTime const& utc,
```

Comment:
Another name, "pre 1972", for the stretchy thingy.  Please pick one word for this and stick to it.